src/RenderAPI/EspRenderScheduler.cc

```diff
@@ -0,0 +1,182 @@
+#include "EspRenderScheduler.hh"
+
+// std
+#include <stdexcept>
+#include <array>
+
+namespace esp
+{
+	EspRenderScheduler::EspRenderScheduler(EspWindow& window, EspDevice& device)
+	: m_window{window}, m_device{device}
+	{
+		recreate_swap_chain();
+		create_command_buffers();
+	}
+
+	EspRenderScheduler::~EspRenderScheduler()
+	{
+		free_command_buffers();
+	}
+
+	VkCommandBuffer EspRenderScheduler::begin_frame()
+	{
+		assert(!m_frame_started && "Cannot call begin_frame while frame already in progress");
+
+		auto result = m_esp_swap_chain->acquire_next_image(&m_current_image_index);
+
+		//window has changed in such a way that is no longer compatible with the swap chain (example: resizing)
+		if (result == VK_ERROR_OUT_OF_DATE_KHR)
+		{
+			recreate_swap_chain();
+			return nullptr;
+		}
+
+		if (result != VK_SUCCESS && result != VK_SUBOPTIMAL_KHR)
+		{
+			ESP_CORE_ERROR("Failed to acquire swap chain image");
+			throw std::runtime_error("Failed to acquire swap chain image");
+		}
+
+		m_frame_started = true;
+
+		auto command_buffer = get_current_command_buffer();
+		VkCommandBufferBeginInfo begin_info{};
+		begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+
+		if (vkBeginCommandBuffer(command_buffer, &begin_info) != VK_SUCCESS)
+		{
+			ESP_CORE_ERROR("Failed to begin recording command buffer");
+			throw std::runtime_error("Failed to begin recording command buffer");
+		}
+
+		return command_buffer;
```

Comment:
Is it intentional to return a copy of command_buffer here?

Reply:
Underneath it's accoually a pointer